File: copyRandomList.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""
#TC:O(n)
#SC:O(n)...We will store a clone mapping for each node entailing linear space complexity with respect to the original list passed to us.
#Ran successfully on Leetcode:Yes
#Algo:Use a hashtable to facilitate the cloning.
#In the hashmap we map original to copy of the nodes
#In create node function we check if the copy we are making already exists in the map else, we create a copy.

class Solution:
    def copyRandomList(self, head):
        self.nlist={None:None} #old_node:new_node
        
        def create_node(node):
            if node in self.nlist:
                return self.nlist[node]
            else:
                c_node=Node(node.val)
                self.nlist[node]=c_node
                return c_node
        
        node=head
        while node:
            new_node=create_node(node)
            new_node.next=create_node(node.next)
            new_node.random=create_node(node.random)
            node=node.next
                
        return self.nlist[head]  
        

File: CopyRandomList.java
// Time Complexity : The time complexity is O(n) where n is the number of nodes
// Space Complexity : The space complexity is O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {

        if(head == null){
            return null;
        }

        Node cur = head;

        // create a clone and append it to its original
        while(cur != null){

            Node copy = new Node(cur.val);

            copy.next = cur.next;
            cur.next = copy;

            cur = copy.next;

        }

        cur = head;

        // create clones random pointer
        while(cur != null){

            if(cur.random != null){
                cur.next.random = cur.random.next;
            }
            cur = cur.next.next;
        }

        Node actualHead = head;
        Node newHead = head.next;

        Node oldCur = head;
        Node newCur = head.next;

        // separate original and clone
        while(oldCur != null){

            oldCur.next = newCur.next;
            if(newCur.next != null){
                newCur.next = newCur.next.next;
            }

            oldCur = oldCur.next;
            newCur = newCur.next;
        }

        return newHead;


    }
}